File: main/botclean/BotClean.py
#!/usr/bin/env python

from itertools import product

def get_pair():
    return map(int, raw_input().split())

bi, bj = get_pair() # bot coordinates
m = n = 5
a = [raw_input().strip() for i in xrange(m)]
assert all(len(row) == n for row in a)

min_distance = m + n
di = dj = None      # dirt coordinates

for i, j in product(xrange(m), xrange(n)):
    if a[i][j] == 'd':
        distance = abs(i - bi) + abs(j - bj)

        if distance < min_distance:
            min_distance = distance
            di = i
            dj = j

if di is None or dj is None:
    raise ValueError('no dirty cells found')

if bj < dj:
    print 'RIGHT'
elif bj > dj:
    print 'LEFT'
elif bi < di:
    print 'DOWN'
elif bi > di:
    print 'UP'
else:
    print 'CLEAN'

File: main/botclean/botclean.rb
#!/usr/bin/env ruby
$VERBOSE = true

SIZE = 5
RANGE = 0...SIZE

bi, bj = gets.split.map(&:to_i)
raise 'bot outside board' unless RANGE.include?(bi) && RANGE.include?(bj)

di = dj = nil
best_distance = Float::INFINITY

RANGE.each do |i|
  row = gets.strip
  raise 'wrong row length' if row.length != SIZE

  row.each_char.with_index.select { |e, _| e == 'd' }.each do |_, j|
    distance = (bi - i).abs + (bj - j).abs # taxicab metric
    next unless distance < best_distance

    best_distance = distance
    di = i
    dj = j
  end
end

raise 'board is already clean!' unless di

if bi < di
  puts 'DOWN'
elsif bi > di
  puts 'UP'
elsif bj < dj
  puts 'RIGHT'
elsif bj > dj
  puts 'LEFT'
else
  puts 'CLEAN'
end
